### PR TITLE
Experimental Table Tree Checkboxes

### PR DESCRIPTION
**Pull request by matheusps**
#### What is the purpose of this pull request?
As the title says

#### What problem is this solving?
Table Tree may support checkboxes.

#### How should this be manually tested?
Clone the repo and `yarn && yarn start`

#### Screenshots or example usage
example

#### Types of changes
- [ ] Bug fix (a non-breaking change which fixes an issue)
- [x] New feature (a non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Requires change to documentation, which has been updated accordingly.

### CODE REVIEW

File: react/components/EXPERIMENTAL_TableTree/hooks/useCheckboxesContext.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import { useContext } from 'react'
+
+import CheckboxesContext from '../checkboxContext'
+
+const useCheckboxesContext = () => {
```
Comment (estacioneto) [inline react/components/EXPERIMENTAL_TableTree/hooks/useCheckboxesContext.ts]: Is this file really needed?

Comment (matheusps) [reply react/components/EXPERIMENTAL_TableTree/hooks/useCheckboxesContext.ts]: Not really, reduced to a single `contexts` file

File: react/components/EXPERIMENTAL_Table/contexts.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  if (!context) {
+    throw OUT_OF_SCOPE_COMPOSITES_ERROR
+  }
+  return context
+}
+
+export const useBulkContext = () => {
+  const context = useContext(BulkContext)
+  return context
+}
```
Comment (vitoriaHeliane) [inline react/components/EXPERIMENTAL_Table/contexts.ts]:
```suggestion
export const useBulkContext = () => useContext(BulkContext)
```

File: react/components/EXPERIMENTAL_Table/typings.d.ts [hunk 3]
```diff
@@ -31,33 +31,11 @@ type TableProps = {
...
-  setAllLinesSelected?: (allLinesSelected: boolean) => void
-  selectAllVisibleRows?: () => void
-}
-
 type TableState = {
   visibleColumns?: Array<Column>
   columns?: Array<Column>
   visibleColumns?: Array<Column>
-  items?: Array<Object>
+  items?: Array<Object> | any
```
Comment (vitoriaHeliane) [inline react/components/EXPERIMENTAL_Table/typings.d.ts]: any? :cry:

File: react/components/EXPERIMENTAL_TableTree/contexts.tsx [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+import { checkboxesHookReturn } from './hooks/useTableTreeCheckboxes'
+
+export const CheckboxesContext = createContext<Partial<checkboxesHookReturn>>(
+  {}
+)
+
+export const useCheckboxesContext = () => {
+  const context = useContext(CheckboxesContext)
+  return context
+}
```
Comment (vitoriaHeliane) [inline react/components/EXPERIMENTAL_TableTree/contexts.tsx]:
```suggestion
export const useCheckboxesContext = () => useContext(CheckboxesContext)
```

File: react/components/EXPERIMENTAL_Table/typings.d.ts (hunk 3 above)
Comment (matheusps) [reply react/components/EXPERIMENTAL_Table/typings.d.ts]: Yep! For reusability proposes. A Table can receive items of any kind. Since this PR it's already big, I'll do another one that uses generics instead of simply allowing `any`.

File: react/components/EXPERIMENTAL_TableTree/hooks/useTableTreeCheckboxes.tsx [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+          ? state.filter(row => row.id !== item.id)
+          : [...state, item]
+      } else {
+        const flatItems = getFlat(item)
+        return state.some(row => row.id === item.id)
+          ? state.filter(row => !idsFrom(flatItems).includes(row.id))
+          : [...state, ...flatItems].reduce(
+              (acc, item) =>
+                idsFrom(acc).includes(item.id) ? acc : [...acc, item],
+              []
```
Comment (analuizamtg) [inline react/components/EXPERIMENTAL_TableTree/hooks/useTableTreeCheckboxes.tsx]: :developers:

File: react/components/EXPERIMENTAL_TableTree/README.md [hunk 6]
```diff
@@ -32,9 +34,19 @@ function ToolbarExample() {
   const [inputValue, setInputValue] = React.useState('')
   const [displayItems, setDisplayItems] = React.useState(items)
 
+  const onToggle = ({ checkedItems }) => {
+    console.log(checkedItems)
```
Comment (mariana-mendes) [inline react/components/EXPERIMENTAL_TableTree/README.md]: :thinking:

File: react/components/EXPERIMENTAL_TableTree/Tree/CellPrefix.tsx [hunk 7]
```diff
@@ -1,45 +1,35 @@
...
-  return (
-    <ButtonWithIcon
-      size="small"
-      onClick={onClick}
-      icon={icon}
-      variation="tertiary"
-    />
-  )
-}
+const PREFIX_WIDTH = 64
```
Comment (rerissondaniel) [inline react/components/EXPERIMENTAL_TableTree/Tree/CellPrefix.tsx]: Would be great if this was somehow customizable, I think this could easily vary between clients

File: react/components/EXPERIMENTAL_TableTree/index.tsx [hunk 8]
```diff
@@ -1,38 +1,60 @@
...
+          </DataTable>
+        </TableContainer>
+      </CheckboxesContext.Provider>
+    </TableContext.Provider>
   )
 }
 
-type Props = InferProps<typeof tablePropTypes>
+const itemTreeShape = PropTypes.shape({
+  children: PropTypes.any,
```
Comment (rerissondaniel) [inline react/components/EXPERIMENTAL_TableTree/index.tsx]:
```suggestion
  children: PropTypes.node,
``` ?

File: react/components/EXPERIMENTAL_TableTree/hooks/useTableTreeCheckboxes.tsx [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+}
+
+function getFlat(tree: ItemTree, arr: Array<ItemTree> = []) {
+  arr.push(tree)
+  if (!tree.children) return arr
+  else {
+    tree.children.forEach(child => getFlat(child, arr))
+    return arr
+  }
+}
```
Comment (mariana-mendes) [inline react/components/EXPERIMENTAL_TableTree/hooks/useTableTreeCheckboxes.tsx]:
```
  arr.push(tree)
  if (tree.children){
    tree.children.forEach(child => getFlat(child, arr))
  }

  return arr

```

File: react/components/EXPERIMENTAL_TableTree/hooks/useTableTreeCheckboxes.tsx (hunk 5 above)
Comment (matheusps) [reply react/components/EXPERIMENTAL_TableTree/hooks/useTableTreeCheckboxes.tsx]: Made some changes, so that code is more readable. 😅

File: react/components/EXPERIMENTAL_TableTree/README.md (hunk 6 above)
Comment (matheusps) [reply react/components/EXPERIMENTAL_TableTree/README.md]: It is just to represent an action that could be done onToggle. (i found `alert` annoying)

File: react/components/EXPERIMENTAL_TableTree/index.tsx (hunk 8 above)
Comment (matheusps) [reply react/components/EXPERIMENTAL_TableTree/index.tsx]: Nope, this is children of an item. Node doesn't fit

File: react/components/EXPERIMENTAL_Table/Checkbox.tsx [hunk 10]
```diff
@@ -14,14 +14,17 @@ const Checkbox: FC<CheckboxProps> = ({ id, onClick, ...props }) => (
 )
 
 export type CheckboxProps = {
-  id: string
-  checked: boolean
-  onClick: Function
+  id?: string
+  onClick?: Function
```
Comment (rerissondaniel) [inline react/components/EXPERIMENTAL_Table/Checkbox.tsx]:
```suggestion
  onClick?: (id: string) => void
```

File: react/components/EXPERIMENTAL_TableTree/hooks/useTableTreeCheckboxes.tsx (hunk 9 above)
Comment (matheusps) [reply react/components/EXPERIMENTAL_TableTree/hooks/useTableTreeCheckboxes.tsx]: Nice catch

Comment (analuizamtg) [reply react/components/EXPERIMENTAL_TableTree/hooks/useTableTreeCheckboxes.tsx]: couldn't this be done with reduce?

File: react/components/EXPERIMENTAL_TableTree/Tree/CellPrefix.tsx (hunk 7 above)
Comment (matheusps) [reply react/components/EXPERIMENTAL_TableTree/Tree/CellPrefix.tsx]: Considering the table usage this will not vary since arrows and checkboxes have fixed widths.

File: react/components/EXPERIMENTAL_TableTree/index.tsx (hunk 8 above)
Comment (rerissondaniel) [reply react/components/EXPERIMENTAL_TableTree/index.tsx]: children shouldn't be anything that is rederizable?

File: react/components/EXPERIMENTAL_TableTree/Tree/CellPrefix.tsx (hunk 7 above)
Comment (matheusps) [reply react/components/EXPERIMENTAL_TableTree/Tree/CellPrefix.tsx]: Also, allowing users to customize it could lead to the `TableTree` looking uneven throughout different admin pages.

Comment (matheusps) [reply react/components/EXPERIMENTAL_TableTree/Tree/CellPrefix.tsx]: Maybe it is something to discuss with @oldufre. A great solution, IMO, is to create a range of different sizes, such as: 
```
INDENTATION: {
  SMALL: 32,
  MEDIUM: 48,
  LARGE: 64
}
```
What do you think about it? 🤔

Comment (rerissondaniel) [reply react/components/EXPERIMENTAL_TableTree/Tree/CellPrefix.tsx]: I think that would be great, I just mentioned because I thought the alignment of the inner arrows was akward, I think they should be aligned to the outer checkboxes

File: react/components/EXPERIMENTAL_TableTree/Tree/index.tsx [hunk 11]
```diff
@@ -62,19 +73,40 @@ const Node: FC<NodeProps> = ({ data, depth }) => {
...
 const Tree: FC = () => {
   const { items } = useTableContext()
-
+  const [collapsedItems, setCollapsedItems] = useState([])
+  const toggleCollapsed = (id: string) => {
+    collapsedItems.includes(id)
+      ? setCollapsedItems(collapsedItems =>
+          collapsedItems.filter(cid => cid !== id)
+        )
+      : setCollapsedItems(collapsedItems => [...collapsedItems, id])
```
Comment (mariana-mendes) [inline react/components/EXPERIMENTAL_TableTree/Tree/index.tsx]: Maybe... 
```
  setCollapsedItems(collapsedItems =>
    collapsedItems.includes(id)  ? 
         collapsedItems.filter(cid => cid !== id) : 
         collapsedItems => [...collapsedItems, id]
  )
```

Comment (matheusps) [reply react/components/EXPERIMENTAL_TableTree/Tree/index.tsx]: Done

File: react/components/EXPERIMENTAL_TableTree/hooks/useTableTreeCheckboxes.tsx (hunk 9 above)
Comment (matheusps) [reply react/components/EXPERIMENTAL_TableTree/hooks/useTableTreeCheckboxes.tsx]: I don't think that `reduce` would give much gain. The code would get complexity, and still the same performance-wise

File: CHANGELOG.md [hunk 12]
```diff
@@ -7,6 +7,14 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
 ## [Unreleased]
 
+### Added
+
+- Checkboxes to `EXPERIMENTAL_TableTree`
+
+### Chagend
```
Comment (kevinch) [inline CHANGELOG.md]:
```suggestion
### Changed
```

File: react/components/EXPERIMENTAL_Table/DataTable/Rows.tsx [hunk 13]
```diff
@@ -15,7 +17,9 @@ const Rows: FC<RowsProps> = ({ cellProps, rowProps }) => {
         }
       : {}
     const isSelected =
-      bulkState && bulkState.selectedRows.some(row => row.id === rowData.id)
+      bulkContext &&
+      bulkContext.bulkState &&
+      bulkContext.bulkState.selectedRows.some(row => row.id === rowData.id)
```
Comment (kevinch) [inline react/components/EXPERIMENTAL_Table/DataTable/Rows.tsx]: No chance of `selectedRows` being `null` or `undefined`?

File: react/components/EXPERIMENTAL_TableTree/README.md (hunk 6 above)
Comment (kevinch) [reply react/components/EXPERIMENTAL_TableTree/README.md]: Fine for a readme indeed.

File: react/components/EXPERIMENTAL_TableTree/Tree/Arrow.tsx [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+    </span>
+  )
+}
+
+export type ArrowProps = {
+  active: boolean
+  onClick: Function
+}
+
+export default Arrow
```
Comment (kevinch) [inline react/components/EXPERIMENTAL_TableTree/Tree/Arrow.tsx]: Could this be rather added to the icon's list?

File: react/components/EXPERIMENTAL_TableTree/__tests__/checkboxesUtils.test.ts [hunk 15]
```diff
@@ -0,0 +1,199 @@
+import { getItemTree, getFlat, getToggledState } from '../hooks/checkboxesUtils'
+
+test('Tree is parsed correctly', () => {
+  expect(getItemTree(exampleItems)).toEqual(exampleTree)
+})
+
+test('Tree is flats correctly', () => {
```
Comment (kevinch) [inline react/components/EXPERIMENTAL_TableTree/__tests__/checkboxesUtils.test.ts]:
```suggestion
test('Tree is flattened correctly', () => {
```
Maybe?

File: react/components/EXPERIMENTAL_TableTree/hooks/checkboxesUtils.ts [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+    return item.children ? bulkCheck(state, item) : [...state, item]
+  }
+}
+
+/**
+ * Represents a tree section on a single array.
+ */
+export function getFlat(tree: ItemTree, arr: Array<ItemTree> = []) {
+  arr.push(tree)
+  tree.children && tree.children.forEach(child => getFlat(child, arr))
```
Comment (estacioneto) [inline react/components/EXPERIMENTAL_TableTree/hooks/checkboxesUtils.ts]: Guards are great when returning a value. For this, it might be too much, don't you think?

File: react/components/EXPERIMENTAL_TableTree/__tests__/checkboxesUtils.test.ts (hunk 15 above)
Comment (matheusps) [reply react/components/EXPERIMENTAL_TableTree/__tests__/checkboxesUtils.test.ts]: Sure

File: react/components/EXPERIMENTAL_TableTree/index.tsx (hunk 8 above)
Comment (matheusps) [reply react/components/EXPERIMENTAL_TableTree/index.tsx]: I share your concern since this can cause confusion with the `children` prop of react components. That's why now I added a prop to allow the user to choose the key that will contain each item children.

File: react/components/EXPERIMENTAL_Table/DataTable/Rows.tsx (hunk 13 above)
Comment (matheusps) [reply react/components/EXPERIMENTAL_Table/DataTable/Rows.tsx]: Nope, since `bulkState` initializes `selectedRows` with an empty array.

File: react/components/EXPERIMENTAL_TableTree/Tree/Arrow.tsx (hunk 14 above)
Comment (matheusps) [reply react/components/EXPERIMENTAL_TableTree/Tree/Arrow.tsx]: Nope. This component is a Button that toggles collapse on each row and uses two icons (arrow up and down). The name was changed to `CollapseToggle` to avoid confusion.

File: react/components/EXPERIMENTAL_Table/hooks/useTableBulkActions.tsx [hunk 17]
```diff
@@ -233,4 +233,26 @@ type hookReturn = {
   selectAllVisibleRows?: () => void
 }
 
+export type BulkedItem = {
+  id: number
```
Comment (rerissondaniel) [inline react/components/EXPERIMENTAL_Table/hooks/useTableBulkActions.tsx]: I think this should be a string... All of our graphql ids are string

File: react/components/EXPERIMENTAL_TableTree/hooks/checkboxesUtils.ts [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+      ...item,
+      id: `${path}.${index}`,
+      [childsKey]: parsedChilden,
+    }
+  }
+  return { id: 'root', [childsKey]: items.map((item, i) => parseTree(item, i)) }
+}
+
+/** Compares ids  */
+export const equalsId = (item: any) => (candidate: any) =>
```
Comment (arthurepc) [inline react/components/EXPERIMENTAL_TableTree/hooks/checkboxesUtils.ts]: Do you already have a type for `item` and `candidate`?

File: react/components/EXPERIMENTAL_TableTree/hooks/useTableTreeCheckboxes.tsx [hunk 19]
```diff
@@ -0,0 +1,138 @@
...
+    [checkedItems]
+  )
+
+  useEffect(() => {
+    onToggle({ checkedItems })
+  }, [toggle])
+
+  useEffect(() => {
+    const shake = (tree: ItemTree) => {
+      const childs = tree[childsKey] as Array<ItemTree>
```
Comment (arthurepc) [inline react/components/EXPERIMENTAL_TableTree/hooks/useTableTreeCheckboxes.tsx]: ⛏ I think that this variable should be named `children`, what do you think?

File: react/components/EXPERIMENTAL_TableTree/Tree/CollapseToggle.tsx [hunk 20]
```diff
@@ -0,0 +1,26 @@
...
+        onClick={onClick}
+        icon={icon}
+        variation="tertiary"
+      />
+    </span>
+  )
+}
+
+export type CollapseToggleProps = {
+  active: boolean
```
Comment (rerissondaniel) [inline react/components/EXPERIMENTAL_TableTree/Tree/CollapseToggle.tsx]:
```suggestion
  collapsed: boolean
```
Just a thought, active is also fine

File: react/components/EXPERIMENTAL_TableTree/Tree/TreeHeadings.tsx [hunk 21]
```diff
@@ -0,0 +1,60 @@
...
+  return (
+    <Row {...rowProps} height={TABLE_HEADER_HEIGHT}>
+      {visibleColumns.map((headerData: Column, headerIndex: number) => {
+        const { headerRender, title, width } = headerData
+        const content = headerRender ? headerRender({ headerData }) : title
+        return (
+          <Row.Cell
+            {...cellProps}
+            className="bt"
+            key={`heading-${uuid()}`}
```
Comment (rerissondaniel) [inline react/components/EXPERIMENTAL_TableTree/Tree/TreeHeadings.tsx]: This uuid is great, didn't wonder something like that existed

File: react/components/EXPERIMENTAL_TableTree/hooks/useTableTreeCheckboxes.tsx (hunk 19 above)
Comment (matheusps) [reply react/components/EXPERIMENTAL_TableTree/hooks/useTableTreeCheckboxes.tsx]: I was using `children`, but found that could mess-up the head a little, since react use the same naming for children of components.

Comment (matheusps) [reply react/components/EXPERIMENTAL_TableTree/hooks/useTableTreeCheckboxes.tsx]: What do you think about naming this after `childNodes`, since we are on a `tree`? (Also, `childsKey` after `nodesKey`). 🤔

File: react/components/EXPERIMENTAL_TableTree/Tree/CollapseToggle.tsx (hunk 20 above)
Comment (matheusps) [reply react/components/EXPERIMENTAL_TableTree/Tree/CollapseToggle.tsx]: collapsed is better, thanks

File: react/components/EXPERIMENTAL_TableTree/Tree/TreeHeadings.tsx (hunk 21 above)
Comment (matheusps) [reply react/components/EXPERIMENTAL_TableTree/Tree/TreeHeadings.tsx]: Yeah! It is awesome

File: react/components/EXPERIMENTAL_TableTree/hooks/useTableTreeCheckboxes.tsx (hunk 19 above)
Comment (arthurepc) [reply react/components/EXPERIMENTAL_TableTree/hooks/useTableTreeCheckboxes.tsx]: Oh, you have a point. I agree with `childNodes`!

File: react/components/EXPERIMENTAL_TableTree/hooks/checkboxesUtils.ts (hunk 18 above)
Comment (matheusps) [reply react/components/EXPERIMENTAL_TableTree/hooks/checkboxesUtils.ts]: Changed a little, now supports comparison of a given prop. (I'm using `unknown` because I want it to be flexible)

File: react/components/EXPERIMENTAL_Table/hooks/useTableBulkActions.tsx (hunk 17 above)
Comment (matheusps) [reply react/components/EXPERIMENTAL_Table/hooks/useTableBulkActions.tsx]: I'm using a prop called `unicityKey` to deal with which property makes each object unique.

File: react/components/EXPERIMENTAL_Table/hooks/useTableBulkActions.tsx [hunk 22]
```diff
@@ -176,19 +179,22 @@ function reducer(state: BulkState, action: Action) {
...
+      } = action
       return state.selectedRows.some(
-        selectedRow => selectedRow.id === action.row.id
+        selectedRow => selectedRow[unicityKey] === row[unicityKey]
       )
         ? {
             ...state,
             selectedRows: state.selectedRows.filter(
-              row => row.id !== action.row.id
+              r => r[unicityKey] !== row[unicityKey]
```
Comment (kevinch) [inline react/components/EXPERIMENTAL_Table/hooks/useTableBulkActions.tsx]: Nitpick: keep the `row` as variable name instead of something less readable like `r`.

File: react/components/EXPERIMENTAL_Table/index.tsx [hunk 23]
```diff
@@ -12,43 +12,41 @@ import { TableContainer, Thead } from './Styled'
 import DataTable from './DataTable'
 import BulkActions from './BulkActions'
 
-const Table: FC<TableProps> & TableComposites = ({
+const Table: FC<Props> & TableComposites = ({
   children,
   state,
   bulk,
+  unicityKey,
```
Comment (kevinch) [inline react/components/EXPERIMENTAL_Table/index.tsx]: What is this?

File: react/components/EXPERIMENTAL_TableTree/contexts.tsx [hunk 24]
```diff
@@ -0,0 +1,22 @@
+import React, { createContext, useContext, FC } from 'react'
+import { checkboxesHookReturn } from './hooks/useTableTreeCheckboxes'
+import { TreeProps, TreeState } from './index'
+import useTreeState from './hooks/useTreeState'
+
+export const CheckboxesContext = createContext<Partial<checkboxesHookReturn>>(
+  {}
```
Comment (kevinch) [inline react/components/EXPERIMENTAL_TableTree/contexts.tsx]: Normal?

Comment (matheusps) [reply react/components/EXPERIMENTAL_TableTree/contexts.tsx]: Yes, `prettier` stuff

File: react/components/EXPERIMENTAL_Table/hooks/useTableBulkActions.tsx (hunk 22 above)
Comment (matheusps) [reply react/components/EXPERIMENTAL_Table/hooks/useTableBulkActions.tsx]: Oops, done. Thanks

File: react/components/EXPERIMENTAL_TableTree/hooks/checkboxesUtils.ts [hunk 25]
```diff
@@ -0,0 +1,57 @@
...
+ * @param state
+ * @param item
+ */
+export function getToggledState(
+  state: Array<Item>,
+  item: Item,
+  nodesKey: string = 'children',
+  unicityKey: string = 'id'
+): Array<Item> {
+  const equalsKey = eqProp(unicityKey)
```
Comment (rerissondaniel) [inline react/components/EXPERIMENTAL_TableTree/hooks/checkboxesUtils.ts]: inline this, no need of renaming

File: react/components/EXPERIMENTAL_Table/hooks/useTableBulkActions.tsx [hunk 26]
```diff
@@ -29,14 +31,10 @@ const useTableBulkActions = ({
 
   const hasBulkActions = hasPrimaryBulkAction || hasSecondaryBulkActions
 
-  const bulkedItems = useMemo(() => {
-    return hasBulkActions && items.map((item, i) => ({ id: i, ...item }))
-  }, [items, columns])
-
   const bulkedColumns = useMemo<Array<Column>>(() => {
     const headerRender = () => {
```
[the remark this answers is not in the document]
Comment (matheusps) [reply react/components/EXPERIMENTAL_Table/hooks/useTableBulkActions.tsx]: I'll keep that in mind and apply into another PR

File: react/components/EXPERIMENTAL_Table/hooks/useTableBulkActions.tsx [hunk 27]
```diff
@@ -29,14 +31,10 @@ const useTableBulkActions = ({
...
       const selectedRowsLength = bulkState.selectedRows.length
-      const itemsLength = bulkedItems.length
+      const itemsLength = items.length
```
[the remark this answers is not in the document]
Comment (matheusps) [reply react/components/EXPERIMENTAL_Table/hooks/useTableBulkActions.tsx]: You're right. There are a lot of nitpicks that I would like to change. I'll do those on a dedicated PR.

**Comment by TasGuerciMaia** [comment]
When there is a row selected the hover on the caret doesn't work well.

**Comment by TasGuerciMaia** [comment]
We agreed to leave the selected like it is on this PR and as we correct the color overlays in the future we get back to this issue.
